test/test_overrides.py

```diff
@@ -44,12 +58,435 @@ def gemm_diag(mat1, mat2, out=None):
 
     return DiagonalTensor(mat1._N, mat1._i * mat2._i)
 
-
-
 t1 = DiagonalTensor(5, 1)
 t2 = DiagonalTensor(5, 2)
+t3 = DiagonalTensor(5, 1)
 
 print(gemm_diag(t1, t2))
 print(torch.gemm(t1, t2))
 print(torch.gemm(t1.tensor(), t2.tensor()))
 print(torch.mm(t1.tensor(), t2.tensor()))
+
+class TestOverride(unittest.TestCase):
+    def test_gemm_equality(self):
+        self.assertEqual(t1, t3)
+
+    def test_gemm_diag(self):
+        self.assertEqual(torch.gemm(t1, t2), t2)
+
+
+requires_torch_function = pytest.mark.skipif(
+    not TORCH_FUNCTION_ENABLED,
+    reason="__torch_function__ dispatch not enabled.")
+
+def _return_not_implemented(self, *args, **kwargs):
+    return NotImplemented
+
+
+# need to define this at the top level to test pickling
+@torch_function_dispatch(lambda tensor: (tensor,))
+def dispatched_one_arg(tensor):
+    """Docstring."""
+    return 'original'
+
+
+@torch_function_dispatch(lambda tensor1, tensor2: (tensor1, tensor2))
+def dispatched_two_arg(tensor1, tensor2):
+    """Docstring."""
+    return 'original'
+
+
+class TestGetImplementingArgs(object):
+
+    def test_tensor(self):
+        tensor = torch.tensor(1)
+
+        args = _get_implementing_args([tensor])
```

Comment:
This is not yet defined. I think it's something like:
```
from torch._overrides import get_overloaded_types_and_args

def _get_implementing_args(args):
    return get_overloaded_types_and_args(args)[0]
```

Have a careful look at https://github.com/numpy/numpy/commit/7104a3e9fe989f43f20765d740cf22dabae563d4 and https://github.com/numpy/numpy/blob/master/numpy/core/src/multiarray/arrayfunction_override.c#L330, that should allow you to figure it out I think.

If it's not easy, we may go straight to implementing this in C++ and ignore the failing tests till then.